<h1>
  <span style="color:#0d8dfc;">ACTIVITY 2:</span>
  <span>UK ACCIDENT DATA ANALYTICS PROJECT</span>
</h1>

<hr>

<h1> ANALYST: RHEALYN VASQUEZ </h1>

In [161]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from scipy.stats import f_oneway
import warnings
warnings.filterwarnings('ignore')

In [162]:
uk_accidents = pd.read_csv('datasets\\uk_accident_data.csv')

In [163]:
#this shows the 5 first row
uk_accidents.head()

Index Accident_Severity Accident Date   Latitude  \
0  200701BS64157           Serious    05/06/2019  51.506187   
1  200701BS65737           Serious    02/07/2019  51.495029   
2  200701BS66127           Serious    26/08/2019  51.517715   
3  200701BS66128           Serious    16/08/2019  51.495478   
4  200701BS66837            Slight    03/09/2019  51.488576   

              Light_Conditions           District Area  Longitude  \
0        Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                     Daylight  Kensington and Chelsea  -0.173647   
2  Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                     Daylight  Kensington and Chelsea  -0.202731   
4        Darkness - lights lit  Kensington and Chelsea  -0.192487   

   Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                     1                   2                     Dry   
1                     1                   2             Wet or damp   
2                     1                   3                     Dry   
3                     1                   4                     Dry   
4                     1                   2                     Dry   

            Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0  Single carriageway               Urban     Fine no high winds   
1  Single carriageway               Urban  Raining no high winds   
2                 NaN               Urban                    NaN   
3  Single carriageway               Urban     Fine no high winds   
4                 NaN               Urban                    NaN   

                           Vehicle_Type  
0                                   Car  
1                                   Car  
2                 Taxi/Private hire car  
3  Bus or coach (17 or more pass seats)  
4                         Other vehicle

In [164]:
#this shows quick summary of the DataFrame
uk_accidents.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 660679 entries, 0 to 660678
Data columns (total 14 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   Index                    660679 non-null  object 
 1   Accident_Severity        660679 non-null  object 
 2   Accident Date            660679 non-null  object 
 3   Latitude                 660654 non-null  float64
 4   Light_Conditions         660679 non-null  object 
 5   District Area            660679 non-null  object 
 6   Longitude                660653 non-null  float64
 7   Number_of_Casualties     660679 non-null  int64  
 8   Number_of_Vehicles       660679 non-null  int64  
 9   Road_Surface_Conditions  659953 non-null  object 
 10  Road_Type                656159 non-null  object 
 11  Urban_or_Rural_Area      660664 non-null  object 
 12  Weather_Conditions       646551 non-null  object 
 13  Vehicle_Type             660679 non-null  object 
dtypes: f

In [165]:
# this shows quick statistics about the data
uk_accidents.describe()

Latitude      Longitude  Number_of_Casualties  Number_of_Vehicles
count  660654.000000  660653.000000         660679.000000       660679.000000
mean       52.553866      -1.431210              1.357040            1.831255
std         1.406922       1.383330              0.824847            0.715269
min        49.914430      -7.516225              1.000000            1.000000
25%        51.490690      -2.332291              1.000000            1.000000
50%        52.315641      -1.411667              1.000000            2.000000
75%        53.453452      -0.232869              1.000000            2.000000
max        60.757544       1.762010             68.000000           32.000000

In [166]:
uk_accidents.isnull().sum()

Index                          0
Accident_Severity              0
Accident Date                  0
Latitude                      25
Light_Conditions               0
District Area                  0
Longitude                     26
Number_of_Casualties           0
Number_of_Vehicles             0
Road_Surface_Conditions      726
Road_Type                   4520
Urban_or_Rural_Area           15
Weather_Conditions         14128
Vehicle_Type                   0
dtype: int64

<h1> FILLING MISSING VALUES/NULL VALUES </h1>

In [167]:
#mean - to handle missing values in numeric columns
for num in ['Latitude', 'Longitude']:
    uk_accidents[num] = uk_accidents[num].fillna(uk_accidents[num].mean())
    
#mode - to handle missing values in categorical columns
for cat in ['Road_Surface_Conditions', 'Urban_or_Rural_Area', 'Weather_Conditions']:
    uk_accidents[cat] = uk_accidents[cat].fillna(uk_accidents[cat].mode()[0])

uk_accidents['Road_Type'] = uk_accidents['Road_Type'].fillna('Unknown')

In [168]:
uk_accidents.isnull().sum()

Index                      0
Accident_Severity          0
Accident Date              0
Latitude                   0
Light_Conditions           0
District Area              0
Longitude                  0
Number_of_Casualties       0
Number_of_Vehicles         0
Road_Surface_Conditions    0
Road_Type                  0
Urban_or_Rural_Area        0
Weather_Conditions         0
Vehicle_Type               0
dtype: int64

In [169]:
uk_accidents['Index'] = uk_accidents['Index'].astype('category')
uk_accidents['Accident_Severity'] = uk_accidents['Accident_Severity'].astype('category')
uk_accidents['Accident Date'] = uk_accidents['Accident Date'].astype('str')
uk_accidents['Accident Date'] = uk_accidents['Accident Date'].str.replace('/', '-')
uk_accidents['Accident Date'] = pd.to_datetime(uk_accidents['Accident Date'], dayfirst = True, errors='coerce')
uk_accidents['Latitude'] = uk_accidents['Latitude'].astype('float')
uk_accidents['Longitude'] = uk_accidents['Longitude'].astype('float')
uk_accidents['Light_Conditions'] = uk_accidents['Light_Conditions'].astype('category')
uk_accidents['District Area'] = uk_accidents['District Area'].astype('category')
uk_accidents['Number_of_Casualties'] = uk_accidents['Number_of_Casualties'].astype('int')
uk_accidents['Number_of_Vehicles'] = uk_accidents['Number_of_Vehicles'].astype('int')
uk_accidents['Road_Surface_Conditions'] = uk_accidents['Road_Surface_Conditions'].astype('category')
uk_accidents['Road_Type'] = uk_accidents['Road_Type'].astype('category')
uk_accidents['Urban_or_Rural_Area'] = uk_accidents['Urban_or_Rural_Area'].astype('category')
uk_accidents['Weather_Conditions'] = uk_accidents['Weather_Conditions'].astype('category')
uk_accidents['Vehicle_Type'] = uk_accidents['Vehicle_Type'].astype('category')

In [170]:
uk_accidents.dtypes

Index                            category
Accident_Severity                category
Accident Date              datetime64[ns]
Latitude                          float64
Light_Conditions                 category
District Area                    category
Longitude                         float64
Number_of_Casualties                int64
Number_of_Vehicles                  int64
Road_Surface_Conditions          category
Road_Type                        category
Urban_or_Rural_Area              category
Weather_Conditions               category
Vehicle_Type                     category
dtype: object

<h1> EXTRACTING DATE INFORMATION USING PANDAS DATE TIME </h1>

In [171]:
uk_accidents['Year'] = uk_accidents['Accident Date'].dt.year
uk_accidents['Month'] = uk_accidents['Accident Date'].dt.month
uk_accidents['Month_Name'] = uk_accidents['Accident Date'].dt.month_name()
uk_accidents['Date'] = uk_accidents['Accident Date'].dt.day
uk_accidents['Day'] = uk_accidents['Accident Date'].dt.day_name()
uk_accidents['Week'] = uk_accidents['Accident Date'].dt.isocalendar().week

In [172]:
uk_accidents[['Accident Date', 'Year', 'Month', 'Month_Name', 'Date', 'Day', 'Week']].head()

Accident Date  Year  Month Month_Name  Date        Day  Week
0    2019-06-05  2019      6       June     5  Wednesday    23
1    2019-07-02  2019      7       July     2    Tuesday    27
2    2019-08-26  2019      8     August    26     Monday    35
3    2019-08-16  2019      8     August    16     Friday    33
4    2019-09-03  2019      9  September     3    Tuesday    36

<h1 style="color:#0d8dfc;"> DATA ANALYSIS</h1>
<hr>

<h1> DATA AGGREGATION</h1>

<h1> 1. Which district has the highest average number of casualties per accident? </h1>

In [173]:
district_casualties = uk_accidents.groupby("District Area")["Number_of_Casualties"].mean()
max_value = district_casualties.max()
district_casualties_percent = (district_casualties / max_value) * 100
result = pd.DataFrame({
    "Average_Casualties": district_casualties,
    "Percentage_of_Max": district_casualties_percent
}).sort_values(by="Average_Casualties", ascending=False)
highest_casualties = result.head(1)
print(highest_casualties)

               Average_Casualties  Percentage_of_Max
District Area                                       
Blaeu Gwent              1.693694              100.0


<h1> INSIGHT 1:</h1>
<p> This shows that the district with the highest average casualties per accident is Blaeu Gwent, where each crash injures about 1.69 or 100% people on average.</p>

<h1> 2. Which road type has the highest average number of vehicles involved? </h1>

In [174]:
uk_accidents.groupby("Road_Type")["Number_of_Vehicles"].mean().sort_values(ascending=False)

Road_Type
Dual carriageway      2.009766
Slip road             1.917483
Roundabout            1.901323
Single carriageway    1.795911
Unknown               1.674115
One way street        1.585441
Name: Number_of_Vehicles, dtype: float64

<h1> INSIGHT 2: </h1>
<p> This shows that crashes on dual carriageways usually involve the most vehicles, while one-way streets involve the fewest. </p>

<h1> 3. Which weather condition is linked with the highest average accident severity? </h1>

In [175]:
uk_accidents['Accident_Severity'] = uk_accidents['Accident_Severity'].cat.codes + 1
uk_accidents.groupby("Weather_Conditions")['Accident_Severity'].mean().sort_values(ascending=False)

Weather_Conditions
Snowing no high winds    2.897884
Other                    2.875743
Snowing + high winds     2.870056
Raining no high winds    2.859918
Raining + high winds     2.838690
Fine no high winds       2.836081
Fog or mist              2.816610
Fine + high winds        2.813538
Name: Accident_Severity, dtype: float64

<h1> INSIGHT 3: </h1>
<p> This shows that the accidents during snowy conditions or unusual weather “Other” tend to be more severe on average, while fine or foggy weather sees slightly lower severity.</p>

<h1> 4. How does the average number of casualties from accidents differ between urban and rural areas? </h1>

In [176]:
uk_accidents.groupby("Urban_or_Rural_Area")["Number_of_Casualties"].mean().sort_values(ascending=False)

Urban_or_Rural_Area
Rural          1.479204
Urban          1.287807
Unallocated    1.181818
Name: Number_of_Casualties, dtype: float64

<h1> INSIGHT 4: </h1>
<p> This shows that the accidents in rural areas have a higher average number of casualties (1.48) compared to urban areas (1.29), with unallocated areas having the lowest average (1.18). This suggests that rural accidents tend to be more severe, possibly due to higher speeds or less safe road conditions.
</p>

<h1> 5. In which light conditions do accidents typically have the highest severity?</h1>

In [177]:
uk_accidents.groupby("Light_Conditions")["Accident_Severity"].mean().sort_values(ascending=False)

Light_Conditions
Darkness - lighting unknown    2.856570
Daylight                       2.853756
Darkness - lights lit          2.823327
Darkness - lights unlit        2.823044
Darkness - no lighting         2.722253
Name: Accident_Severity, dtype: float64

<h1> INSIGHT 5:</h1>
<p>This shows that while darkness with unknown lighting(2.86) is linked to the most severe accidents on average, severity differences between lighting conditions are relatively small.</p>

<hr> 
<h1> Correlation </h1>

<h1> 6. How is the number of casualties related to the number of vehicles involved in an accident? </h1>

In [178]:
uk_accidents["Number_of_Casualties"].corr(uk_accidents["Number_of_Vehicles"])

np.float64(0.2288888612692756)

<h1> INSIGHT 6: </h1>
<p>A positive correlation of 0.23 indicates that accidents with more vehicles tend to have slightly more casualties, though the relationship is weak.
</p>

<h1>7. Does the latitude affect how severe accidents are? </h1>

In [179]:
uk_accidents["Latitude"].corr(uk_accidents["Accident_Severity"])

np.float64(-0.026468995471298485)

<h1>INSIGHT 7:</h1>
<p>This shows that there is no correlation between latitude and accident severity, meaning that where an accident occurs north or south does not affect how severe it is.
</p>

<h1> 8. Has the number of casualties changed over time?</h1>

In [180]:
uk_accidents["Accident Date"] = pd.to_datetime(uk_accidents["Accident Date"], errors='coerce')
uk_accidents = uk_accidents.dropna(subset=["Accident Date"])
uk_accidents["Date_num"] = uk_accidents["Accident Date"].map(pd.Timestamp.toordinal)
uk_accidents["Date_num"].corr(uk_accidents["Number_of_Casualties"])

np.float64(-0.0005681761764865083)

<h1> INSIGHT 8:</h1>
<p>This shows that there is no correlation between accident date and the number of casualties, indicating that the number of casualties has not changed significantly over time.
</p>

<hr>
<h1>DESCRIPTIVE</h1>

<h1>9. Which types of roads are most commonly involved in accidents? </h1>

In [181]:
road_counts = uk_accidents["Road_Type"].value_counts().sort_values(ascending=False)
road_counts

Road_Type
Single carriageway    492143
Dual carriageway       99424
Roundabout             43992
One way street         13559
Slip road               7041
Unknown                 4520
Name: count, dtype: int64

<h1> INSIGHT 9: </h1>
<p> This suggests that single carriageways are the most accident-prone, accounting for about 73.8% of all accidents. This may be due to higher traffic density, limited overtaking opportunities, or narrower lanes.</p>

<h1> 10. How are accidents distributed between urban and rural areas?</h1>

In [182]:
area_counts = uk_accidents["Urban_or_Rural_Area"].value_counts()
area_counts

Urban_or_Rural_Area
Urban          421678
Rural          238990
Unallocated        11
Name: count, dtype: int64

<h1>INSIGHT 10: </h1>
<p>This indicates that urban environments account for about 63.8% of all accidents, likely due to higher traffic density and more complex road networks. In contrast, rural areas account for 36.2%, and only a negligible share (0.0%) is unallocated.</p>

<hr>
<h1>(Categorical → Numerical) </h1>

<h1>11. Do different road types have different average casualties?</h1>

In [183]:
groups = [g["Number_of_Casualties"].values for _, g in uk_accidents.groupby("Road_Type")]
f_oneway(*groups)

F_onewayResult(statistic=np.float64(667.8581812346322), pvalue=np.float64(0.0))

<h1>INSIGHT 11:</h1>
<p>This shows that the average number of casualties differs significantly between road types, with some roads, like single carriageways, tending to have more casualties than others.
</p>

<h1> 12.  Does light condition affect the number of vehicles involved? </h1>

In [184]:
groups = [g["Number_of_Vehicles"].values for _, g in uk_accidents.groupby("Light_Conditions")]
f_oneway(*groups)

F_onewayResult(statistic=np.float64(1490.4430820056236), pvalue=np.float64(0.0))

<h1> INSIGHT 12:</h1>
<p>This shows that accidents involving more vehicles tend to have slightly more casualties, but the relationship is weak.
</p>

<h1> 13. Do different districts show different casualty numbers? </h1>

In [185]:
f_oneway(*[g["Number_of_Casualties"].values for _, g in uk_accidents.groupby("District Area")])

F_onewayResult(statistic=np.float64(24.100343506650468), pvalue=np.float64(0.0))

<h1>INSIGHT 13: </h1>
<p>This shows that some districts have more casualties per accident than others.</p>

<h1>DESCRIPTIVE COMPARISON </h1>

<h1>14. Which vehicle type is linked to the highest average severity?</h1>

In [186]:
uk_accidents.groupby("Vehicle_Type")["Accident_Severity"].mean().sort_values(ascending=False)

Vehicle_Type
Data missing or out of range             3.000000
Ridden horse                             3.000000
Bus or coach (17 or more pass seats)     2.844540
Taxi/Private hire car                    2.843463
Motorcycle 125cc and under               2.842229
Motorcycle 50cc and under                2.841641
Goods 7.5 tonnes mgw and over            2.840931
Car                                      2.840128
Motorcycle over 125cc and up to 500cc    2.840125
Van / Goods 3.5 tonnes mgw or under      2.839344
Other vehicle                            2.839099
Motorcycle over 500cc                    2.838835
Goods over 3.5t. and under 7.5t          2.837434
Agricultural vehicle                     2.833590
Minibus (8 - 16 passenger seats)         2.830972
Pedal cycle                              2.741117
Name: Accident_Severity, dtype: float64

<h1>INSIGHT 14: </h1>
<p>This shows that ridden horses and missing data cases have the highest severity. This suggests that motorcycles and mid-sized goods vehicles are linked to more severe accidents, while pedal cycles are the least severe.
</p>

<h1> 15. Which road surface leads to the highest casualties?</h1>

In [187]:
uk_accidents.groupby("Road_Surface_Conditions")["Number_of_Casualties"].mean().sort_values(ascending=False)

Road_Surface_Conditions
Flood over 3cm. deep    1.495575
Wet or damp             1.411064
Snow                    1.359593
Frost or ice            1.348815
Dry                     1.334545
Name: Number_of_Casualties, dtype: float64

<h1>INSIGHT 15:</h1>
<p>This shows that roads with flood over 3cm deep have the highest average casualties. This indicates that wet, snowy, or icy conditions also increase casualties, while dry roads have the lowest.</p>

<h1>16. What is the trend of average vehicles per accident across different years?</h1>

In [188]:
vehicle_year = uk_accidents.groupby(uk_accidents["Accident Date"].dt.year)["Number_of_Vehicles"].mean()
vehicle_year

Accident Date
2019    1.839310
2020    1.826615
2021    1.826229
2022    1.832273
Name: Number_of_Vehicles, dtype: float64

<h1>INSIGHT 16: </h1>
<p>This shows that the average number of vehicles per accident has remained fairly stable over time, ranging between 1.82 and 1.84, with only slight fluctuations across the years.</p>

<h1>17. Which accidents involved both motorcycles and high-speed roads (dual carriageways)?</h1>

In [189]:
motor_highspeed = uk_accidents[
    (uk_accidents["Vehicle_Type"].str.contains("Motorcycle")) &
    (uk_accidents["Road_Type"] == "Dual carriageway")
]

motor_summary = motor_highspeed["Vehicle_Type"].value_counts().sort_values(ascending=False)
motor_summary


Vehicle_Type
Motorcycle over 500cc                    3930
Motorcycle 125cc and under               2228
Motorcycle over 125cc and up to 500cc    1210
Motorcycle 50cc and under                1085
Data missing or out of range                0
Agricultural vehicle                        0
Car                                         0
Bus or coach (17 or more pass seats)        0
Minibus (8 - 16 passenger seats)            0
Goods over 3.5t. and under 7.5t             0
Goods 7.5 tonnes mgw and over               0
Other vehicle                               0
Pedal cycle                                 0
Ridden horse                                0
Taxi/Private hire car                       0
Van / Goods 3.5 tonnes mgw or under         0
Name: count, dtype: int64

<h1>INSIGHT 17: </h1>
<p>This shows that motorcycles on dual carriageways are involved in accidents, with Motorcycle over 500cc having the highest count, indicating that high-speed roads are riskier for powerful motorcycles.</p>

<h1>18. Are there vehicle types that rarely appear in accidents?</h1>

In [190]:
rare = uk_accidents["Vehicle_Type"].value_counts().nsmallest(5)
rare

Vehicle_Type
Ridden horse                           4
Data missing or out of range           6
Pedal cycle                          197
Agricultural vehicle                1947
Minibus (8 - 16 passenger seats)    1976
Name: count, dtype: int64

<h1>INSIGHT 18: </h1>
<p>This shows that ridden horse and pedal cycle are involved in fewer accidents, indicating these vehicles are less common on accident-prone roads.</p>

<h1> 19. Which light condition occurs least often in accidents?</h1>

In [191]:
uk_accidents["Light_Conditions"].value_counts().nsmallest(1)

Light_Conditions
Darkness - lights unlit    2543
Name: count, dtype: int64

<h1>INSIGHT 19: </h1>
<p>This shows that darkness with lights unlit is the least common accident scenario, suggesting most accidents happen under some level of visibility.</p>

<h1>20. Which road types have the most accidents at night?</h1>

In [192]:
uk_accidents[uk_accidents["Light_Conditions"].str.contains("Darkness")]["Road_Type"].value_counts()

Road_Type
Single carriageway    128878
Dual carriageway       29072
Roundabout             11411
One way street          3595
Slip road               1792
Unknown                 1051
Name: count, dtype: int64

<h1>INSIGHT 20: </h1>
<p>This shows that single carriageways have the most accidents at night (73.7%), followed by dual carriageways (16.6%) and roundabouts (6.5%), indicating that these roads are more prone to nighttime accidents.</p>

<h1> 21. Which year and month recorded the highest number of accidents? </h1>

In [193]:
uk_accidents["Accident Date"] = pd.to_datetime(uk_accidents["Accident Date"])
uk_accidents["Year"] = uk_accidents["Accident Date"].dt.year
uk_accidents["Month"] = uk_accidents["Accident Date"].dt.month_name()
year_month_counts = uk_accidents.groupby(["Year", "Month"]).size().sort_values(ascending=False)
highest_accident_period = year_month_counts.head(1)
print(highest_accident_period)

Year  Month   
2019  November    16559
dtype: int64


<h1> INSIGHT 21:  </h1>
<p>The highest number of accidents occurred in November 2019 (16,559 cases). This spike may be linked to holiday-related travel and gatherings, alongside seasonal factors like shorter days and adverse weather.</p>

<h1>22. Which day of the week in November records the most accidents?</h1>

In [194]:
uk_accidents["Accident Date"] = pd.to_datetime(uk_accidents["Accident Date"])
november_data = uk_accidents[uk_accidents["Accident Date"].dt.month == 11]

november_data["Day_Name"] = november_data["Accident Date"].dt.day_name()
november_counts = november_data.groupby("Day_Name").size().reset_index(name="Accident_Count")

most_accident_day = november_counts.sort_values(by="Accident_Count", ascending=False).head(1)
print(most_accident_day.to_string(index=False))


Day_Name  Accident_Count
Saturday           10301


<h1> INSIGHT 22: </h1>
<p>The analysis shows that Saturday has the highest number of accidents in November, with 10,301 cases. This suggests weekends may be more accident-prone, possibly due to increased leisure travel, social activities, and late-night driving.</p>

<h1> UNIVARIATE: ANALYSIS OF VEHICLE TYPE IN 2022</h1>
<hr>

<h1> 23. Which are the top 3 vehicle types most commonly involved in accidents in 2022?</h1>

In [195]:
accidents_2022 = uk_accidents[uk_accidents["Year"] == 2022]

In [196]:
vehicle_counts_2022 = accidents_2022["Vehicle_Type"].value_counts()
vehicle_percent_2022 = (vehicle_counts_2022 / vehicle_counts_2022.sum() * 100).round(1)

vehicle_table_2022 = pd.DataFrame({
    "Accident_Count": vehicle_counts_2022,
    "Percentage_of_2022": vehicle_percent_2022
})

print(vehicle_table_2022.to_string())


                                       Accident_Count  Percentage_of_2022
Vehicle_Type                                                             
Car                                            111768                77.4
Van / Goods 3.5 tonnes mgw or under              7589                 5.3
Motorcycle over 500cc                            5116                 3.5
Bus or coach (17 or more pass seats)             4531                 3.1
Motorcycle 125cc and under                       3579                 2.5
Goods 7.5 tonnes mgw and over                    2963                 2.1
Taxi/Private hire car                            2838                 2.0
Motorcycle 50cc and under                        1735                 1.2
Motorcycle over 125cc and up to 500cc            1370                 0.9
Goods over 3.5t. and under 7.5t                  1107                 0.8
Other vehicle                                    1096                 0.8
Minibus (8 - 16 passenger seats)      

<h1> INSIGHT 23: </h1>
<p>The analysis shows that the top 3 vehicle types most commonly involved in accidents in 2022 are <strong>cars</strong> (77.4%), followed by <strong>vans under 3.5 tonnes</strong> (5.3%) and <strong>motorcycles over 500cc</strong> (3.5%). 
This indicates that cars are by far the most frequent vehicle involved in accidents, likely due to their prevalence on the roads, while smaller numbers involve larger vehicles and high-powered motorcycles.</p>


<h1> BIVARIATE: VEHICLE TYPE vs ROAD TYPE</h1>
<hr>

<h1>24. On which road types are motorcycles accidents most common?</h1>

In [197]:
motorcycle_accidents = uk_accidents[uk_accidents["Vehicle_Type"].str.contains("Motorcycle")]
motorcycle_road_percent = pd.crosstab(
    motorcycle_accidents["Vehicle_Type"], 
    motorcycle_accidents["Road_Type"], 
    normalize='index'
) * 100
motorcycle_road_percent = motorcycle_road_percent.round(1)
print(motorcycle_road_percent.to_string())

Road_Type                              Dual carriageway  One way street  Roundabout  Single carriageway  Slip road  Unknown
Vehicle_Type                                                                                                               
Motorcycle 125cc and under                         14.6             2.0         6.5                75.4        1.0      0.6
Motorcycle 50cc and under                          14.3             1.7         6.7                75.5        1.2      0.5
Motorcycle over 125cc and up to 500cc              15.8             2.0         6.4                73.9        1.3      0.5
Motorcycle over 500cc                              15.3             1.9         6.8                74.2        1.1      0.6


<h1> INSIGHT 24: </h1>
<p>All motorcycle types have the majority of accidents on <strong>single carriageways</strong> (74–76%), followed by <strong>dual carriageways</strong> (14–16%) and roundabouts (6–7%). Very few accidents occur on one-way streets, slip roads, or unknown road types. 
This highlights that motorcycles are most at risk on rural or high-speed roads.</p>


<h1> MULTIVARIATE: ACCIDENT DATE, WEATHER CONDITIONS AND NUMBER OF CASUALTIES</h1>
<hr>

<h1> 25. Which month and weather condition combination results in the highest average casualties per accident?</h1>

In [198]:
uk_accidents["Accident Date"] = pd.to_datetime(uk_accidents["Accident Date"])
uk_accidents["Month"] = uk_accidents["Accident Date"].dt.month_name()
weather_month_casualties = uk_accidents.groupby(
    ["Month", "Weather_Conditions"]
)["Number_of_Casualties"].mean().reset_index(name="Average_Casualties")
weather_month_casualties = weather_month_casualties.sort_values(by="Average_Casualties", ascending=False)

print(weather_month_casualties.head(5).to_string(index=False))


    Month    Weather_Conditions  Average_Casualties
   August  Snowing + high winds            2.500000
     June           Fog or mist            1.774510
September Snowing no high winds            1.615385
  October           Fog or mist            1.615108
     July           Fog or mist            1.611111


<h1> INSIGHT 25: </h1>
<p>Accidents in <strong>August with "Snowing + high winds"</strong> are the most severe, while fog or mist in June, July, and October, and snow in September reach 64–71% of the maximum severity. Adverse weather clearly increases accident severity.</p>